### ISSUE TITLE
Show scale (i.e. real-life distances) on the map

### ISSUE THREAD

**Issue by mnalis** [streetcomplete/StreetComplete#5765]
**Use case**
Especially when I am adding things to the map (e.g. using Things or Shop Overlay), but also when verifying existing map, it is not always easy to judge how big distances displayed on the map are, yet it is very important to place things on the map e.g. will those two things I just added to the map end up being `5m` apart or `50m` apart? 

See also that missing functionality contributing to other issues: https://github.com/streetcomplete/StreetComplete/issues/5760#issuecomment-2244824030

**Proposed Solution**

It would be great if there was some scale / indication on the map **showing how much an 1 cm on screen is actual distance in reality**. Here is how some other apps (OsmAnd, Commons App and Vespucci) show those (I personally like OsmAnd one the best, but all are usable):

![small_Screenshot_20240723_164844_OsmAnd~](https://github.com/user-attachments/assets/4fc8b2a7-885b-4cc6-87e6-43afac824687) ![small_Screenshot_20240723_164416_Vespucci](https://github.com/user-attachments/assets/57b61349-9904-41cb-aa10-860119a04c05) ![small_Screenshot_20240723_164200_Commons](https://github.com/user-attachments/assets/2e6f5160-1ef2-4951-aeb0-17afc5ecd97f)

**Comment by westnordost** [comment]
When the scale is only helpful when placing things on the map, it means that it should probably only be shown when either placing things is possible or even only when currently placing a thing.

Then again, a scale shown at the bottom of the screen is problematic as it will both be hidden by the bottom sheet and it is potentially very far away from the distance one is interested it, e.g. the distance of some building corner to the crosshair or whatever.

**Comment by mnalis** [comment]
> When the scale is only helpful when placing things on the map

While that is main use case, it might also useful when verifying that existing POIs are placed correctly. But yes, it would be huge improvement even if it was shown only when placing things on the map, as I encounter the problem there majority of the times.

> Then again, a scale shown at the bottom of the screen is problematic as it will both be hidden by the bottom sheet

Placing it at the top (like in the `Commons app` example above) would avoid that particular issue

> [...] and it is potentially very far away from the distance one is interested it, e.g. the distance of some building corner to the crosshair or whatever.

True, but that is not a big issue in practice (in my experience at least). But if you feel it is, and SC decides to go with center-based placement (see https://github.com/streetcomplete/StreetComplete/issues/5760), then there is also an option to do center-based "concentric circles"; like this more complex OsmAnd example:

![small_Screenshot_20240801_021822_OsmAnd~](https://github.com/user-attachments/assets/ed2750b8-1278-42b4-bece-9d46362ef0a5)

(the OsmAnd example shown here contains much extra cruft, like compass directions, heading etc. which are not needed for this use case; it is solely _"concentric circles with distances"_  e.g. _"50m", "100m", "150m", "200m", "250m"_ that I suggest here)

**Comment by Nick-Tallguy** [comment]
Personally I would find it useful to have the scale when adding 'things' - I might find it clutters the screen at other times.

**Comment by peternewman** [comment]
This was previously raised as #2376 .

I'd agree with @mnalis regarding it being relevant when doing existence checks etc too (is it that bench or another one, is it misplaced or missing), particularly in places with fewer other points of reference (e.g. in a park/field). I've repeated my other comments from last time.

>> I was going to open one for this before, I'm not sure I ever found this one though. I was searching for scale.

> If you're answering an existence check for example, then scale is important, is it this bench or not (how close is the next one).

>> It also differs quite significantly from paper maps by

> Also being able to zoom in and out trivially, therefore throwing out any innate knowledge of scale. I got caught out by this on a paper map as they're 1:10k in Luxembourg because the country is so small and I kept overshooting as I was used to 1:25k and therefore roughly where you'd expect things.

_Originally posted by @peternewman in https://github.com/streetcomplete/StreetComplete/issues/2376#issuecomment-966799768_

**Comment by westnordost** [comment]
How would it fit at the top? There's already the stars counter, the menu button, the overlays button, the upload button and the messages button.

**Comment by mnalis** [comment]
> How would it fit at the top?

In that second [OsmAnd example](https://github.com/streetcomplete/StreetComplete/issues/5765#issuecomment-2261705896), ruler is only slightly wider than a button, comfortably fitting between 4 buttons in the same row (with enough space to fit a 5th button). (it's on the bottom row in that OsmAnd example, but it makes no difference)

here is quick copy/paste mockup (just for size comparison, so ignore the colors etc) of that OsmAnd ruler to SC:

![ruler_sc](https://github.com/user-attachments/assets/b131408f-d8c8-42b5-bd63-c4c3ee8387f4)

**Comment by westnordost** [comment]
There can be a button to the left of the upload button (the message button)

**Comment by mnalis** [comment]
Perhaps put it below stars then, or in bottom right? (quick mockups - more margin should be added from existing buttons)

![ruler_sc_TL](https://github.com/user-attachments/assets/c5fe0396-551a-44a0-81c2-1ec3f961fad1) ![ruler_sc_BR](https://github.com/user-attachments/assets/167e6983-b018-43bc-8c35-db5dca2ab8ac)

Or, if if will only be displayed while adding things, ruler could be bigger / placed elsewhere / in some other style (see previous screenshots from other apps earlier in the thread)

**Comment by westnordost** [comment]
I guess below the star counter would work, space-wise. On the bottom right corner, there can be the stop-recording button.
There'd be more space and it would be a little less busy if the scale was vertical, but that is really uncommon, isn't it?

**Comment by mnalis** [comment]
>  There'd be more space and it would be a little less busy if the scale was vertical, but that is really uncommon, isn't it?

Yes, that would be quite uncommon... I've only seen Vespucci do it (see pics in original report), but even it does both it as part of full-width horizontal+vertical scales. Tried mockup, looks somewhat strange, but not unusable:

![ruler_vertical](https://github.com/user-attachments/assets/59d0020b-066b-4f4b-b9bd-74efefd47b00)

**Comment by lgommans** [comment]
The stars counter could also be hidden at times where the ruler is shown in its place (while positioning elements), since stars aren't relevant during the short time this task takes to complete

(For me, the change in GNSS position as I step around gives a good idea of scale or, in urban areas, the buildings are a good reference. Was just browsing issues as this occurred to me.)

**Comment by mnalis** [comment]
**TL;DR:** no, IMHO small static ruler would be better than explicit user-triggered dynamic measurements  (even if those would be more precise) .

---

... and not just because of ease of discoverability, but ease of use - quick glance is always easier than finger gestures (and *hugely more so* if those gestures happen to require using both hands or multiple touches!)

> it would turn out to be less useful than it could be, because it is only shown in one corner of the screen

Not really. As noted in examples earlier, I do use such static rulers in several different apps with different layout, and do not have much problem with it. Note I never use it a geodesic tool -- high-definition precision is **not** the idea, but just getting the approximate estimation of distance (i.e. scale).

IOW, when mapping in SC I'm not really interested if that is `5.9` or `6.1` meters between those two mapped shops/atms/benches/whatever; but I'm very interested to know if that distance is about `half a meter`, about `5 meters` or about `20 meters` or about `100 meters`.  

**Even if it the app would be more precise**, it would be **useless in practice**, unless I were carrying and using a measurement tape (or laser distance measurement tool) all the time.  Sure, app may tell me those two ATMs are exactly `5.12` meters apart, but my eyes/internal navigation will not be able to confirm or that (at best I could say by quick eyesight without significant effort is that it is about 4-6 meters apart, and even [GPS measurements will have errors](https://en.wikipedia.org/wiki/Error_analysis_for_the_Global_Positioning_System))

> What would be more useful to have is the ability to measure the distance between two arbitrary points on the map. But how to trigger this?

Well, those whole idea about having to explicitly manually trigger this ruler display reduces its usability significantly. And, as your suggestions note, if the ruler jumps randomly on _every_ screen touch (which partially resolves that usability issue by makes it very easy to trigger) it will get annoying in those 90% of the cases when you don't absolutely need a ruler (but might still come handy if was always-on).

OsmAnd for example does offer that capability of user-triggered precise measurements **too** (as a _third_ OsmAnd way to measure distances, in addition to the two mentioned before. There if a also a _fourth_ way in OsmAnd, that allows composite ways and is even more powerful, but that is part of complex full `.gpx` editor, so I won't detail it here). 


![small_Screenshot_20241017_205839_OsmAnd~](https://github.com/user-attachments/assets/d88487fb-0fb0-4381-83ac-1b9703086a90)


That third way allows user-triggered measurements both as a  _current-location-to-touch-point_ and as a _multitouch-distance-between-two-points_; but I've rarely if ever used it, as additional complexity and UI overload was not worth it - slight mental calculation of just looking at the ruler and rotating, transposing and multiplying was practically always more than good enough for me. 
(even if we're talking kilometers instead of meters where difference is much more noticable, my instant approximation of _"about 5km more biking over that grass path until we reach asphalt again"_  was always more welcome than _"let use stop for a moment and I can look it up and precisely tell you if it is 4.7 or 5.3km"_. While SC mapping is obviously quite different than dynamic bicycle route planning, the same issue of _"ease of convenience is much more important than **extreme** precision"_ maps surprisingly well to that use case too)

**Comment by mnalis** [comment]
> I think what would somewhat mitigate the thirst for such a UI feature would be to display satellite imagery, but this is another ticket.

Only in (very limited) amount of cases. 

E.g. in that OsmAnd example picture [above](https://github.com/streetcomplete/StreetComplete/issues/5765#issuecomment-2422008693) (i.e. big buildings with multiple amenities, which is one of the significant uses, if not the main one), satellite imagery wouldn't help at all. 

And in many other cases where satellite imagery might help, it often doesn't because the resolutions are usually too low. (I do use satellite imagery both in Vespucci, OsmAnd and SCEE from time to time, but even local Croatian aerial imageries which are better resolution then public Bing/Maxxar are very often insufficient. **And** then there is the dynamic satellite imagery [offset issue](https://wiki.openstreetmap.org/wiki/Offset) with them, dealing with which which pushes them to Vespucci or SCEE territory at least)

**Comment by westnordost** [comment]
with the new arrangement of buttons in v60, below the buttons on the upper right corner might be a reasonable space for this. (PR welcome)

**Comment by mnalis** [comment]
> with the new arrangement of buttons in v60, below the buttons on the upper right corner might be a reasonable space for this.

Something like this? (too complex for me to implement, but a mockup to potentially help anyone doing it) 

![ruler_sc60_topright_big](https://github.com/user-attachments/assets/5129a472-767d-4045-a608-96e797de2ae2)

(style and color are of course better chosen with someone with actual artistic ability :smile_cat: )

**Comment by mnalis** [comment]
From https://github.com/streetcomplete/StreetComplete/issues/6138#issuecomment-2703596559:

> By the way, I contributed to maplibre-compose, a library I hope to migrate to in the future ([#6072](https://github.com/streetcomplete/StreetComplete/issues/6072)) a [scalebar](https://github.com/sargunv/maplibre-compose/pull/232) (fixes [#5765](https://github.com/streetcomplete/StreetComplete/issues/5765)) and a more compact [attribution widget](https://github.com/sargunv/maplibre-compose/pull/248#issuecomment-2591419955) (fixes issues with accidental clicks on the attribution).

Wow, that is great news!

> While a migration might be still somewhat blue sky, maybe in the meantime I can just copy over my work to StreetComplete in the effort to enhance the main UI.

**Yes _please_**, I'm acutely aching for that scalebar!

So if it is going to take some time for a switch to maplibre-compose, it would be great to have at least scalebar backported to StreetComplete sooner rather than later (attribution widget is of course also very nice to clear up a little more screen, but not nearly as operationally useful as scalebar is).
